Kernel API for traversals

pontusmelke opened the pull request:
Switches to Kernel API for
- `NodeProxy:getRelationships` and `NodeProxy:hasRelationships`
- Cypher interpreted runtimes, i.e. `QueryContext:getRelationshipsForIds` and `QueryContext:getRelationshipsForIdsPrimitive`
- Cypher compiled runtime: `Expand(All)`, `Expand(Into)`, and `type(r)`.

Also adds two new concrete helper cursors in the Kernel API, which use RelationshipTraversalCursor and RelationshipGroupCursor to iterate over the relationships of a node, and filter for type and direction. Also add common interface so node-density is possible to abstract away if wanted to.

Replaces https://github.com/neo4j/neo4j/pull/10900

fickludd committed:
Migrate Core API NodeProxy.getRelationships impl to Kernel API

fickludd committed:
Remove test asserting iteration across transactions

fickludd committed:
Helper cursors for relationship selection for dense and sparse nodes

Two new cursors which use interface RelationshipTraversalCursor and RelationshipGroupCursor
to iterate over the relationships of a node, and filter for type and direction. Also add
common interface so node-density is possible to abstract away if wanted to.

fickludd committed:
The day is today

fickludd committed:
Specialize NodeProxy.getRelationships for dense nodes

Also improve traversal & group cursor robustness against concurrent
modification by record loading using RecordLoad.FORCE. Now if a record
is deleted after the cursor got the reference, the whole record is loaded
anyway, and we can continue to the next link in the chain by reading the
now unused record. This works because deleting a record R redirects the previous
and next records in the chain and sets R.inUse = false, but doesn't modify
any other contents of R.

fickludd committed:
Migrate Core API NodeProxy.hasRelationships impl to Kernel API

pontusmelke committed:
Use new API for primitive iteration over relationships

Switch to new API for `getRelationsByIdsPrimitive` and also some code
clean up around wrapping cursors in iterators.

pontusmelke committed:
Use new API for iteration over relationships values

Switch to new API for `getRelationsByIds`.

pontusmelke committed:
Code clean up

pontusmelke committed:
Don't use cursor after close

fickludd committed:
Change RelationshipSelection helper from cursors to iterators

This better fits with current usage patterns in Core API and Cypher
runtimes.

fickludd committed:
Big refactor of relationship selection helpers

Depending on the usage pattern, it is sometimes convenient to use
an iterator returning relationship object of some type, and sometimes
better to use a cursor approach where we do not need to create extra
objects. To achieve this but only have one implementation of the
selection logic itself, we ended up with this rather complex design,
and the bastard method fetchNext with both returns a boolean, and
does a callback to allow object creation.

pontusmelke committed:
Update of underlying helpers

We now have an iterator approach when we want actual objects and a cursor
for the cases where all we want is a reference.

pontusmelke committed:
Some clean up of iterator

thobe commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -1007,77 +1030,94 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
 
-  private def nodeValueIndexCursorToIterator(nodeCursor: NodeValueIndexCursor): Iterator[NodeValue] =  new Iterator[NodeValue] {
-    private var _next: NodeValue = fetchNext()
+  class RelationShipCursorIterator(selectionCursor: RelationshipSelectionCursor) extends RelationshipIterator {
```
A ship full of relations?

thobe reviewed (commented):
Looking at the use of the cursors from #10900 here, they are again used as the basis for constructing iterators, which makes me think that the interface we are exposing through the `RelationshipSelectionCursor` is more powerful than it needs to be.

pontusmelke committed:
Use new API for expand all in compiled runtime

pontusmelke committed:
Use new API for expand into in compiled runtime

pontusmelke committed:
Use new API for type(r)

also clean up and fix some failing tests

pontusmelke committed:
Fixed weird capitalization

pontusmelke commented:
@thobe They are used as a basis for `Iterator[RelationshipValue]` as well as the primitive `RelationshipIterator` in the interpreted runtime and now in the compiled runtime they are used more directly. I think we need to do something about the duplicated logic around getting the right `RelationshipSelection[Cursor/Iterator]` given the denseness, direction etc. As it is now we have basically the same code in three places with some small subtle differences.

pontusmelke committed:
Move all stubs to same package

pontusmelke committed:
Delete CodeGeneratorTest

This was the first ever test we wrote for the compiled runtime, and
it mocks the entire database. It might have been a good idea at the time
but now it has turned into a maintenance nightmare and it doesn't test
stuff that are not tested in numerous other places on _real_ graphs.

pontusmelke committed:
Move logic to create cursors to static method

Move all logic to create the appropriate cursor given denseness,
direction and types to a central place instead of redoing the same
stuff everywhere.

thobe commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/helpers/RelationshipSelections.java:
```diff
@@ -0,0 +1,156 @@
+ */
+package org.neo4j.internal.kernel.api.helpers;
+
+import org.neo4j.graphdb.Direction;
```
We wanted to avoid depending on the graphdb API here, did we not?

thobe replied:
I'd change the design of this class to provide 6 static methods:

* `outgoingIterator(...)`
* `incomingIterator(...)`
* `allIterator(...)`
* `outgoingCursor(...)`
* `incomingCursor(...)`
* `allCursor(...)`

That way there would be no need to depend on `graphdb.Direction`, and if some caller has another encoding of direction it wouldn't need to translate.

thobe reviewed (approved):
Looks to be fitting together pretty nicely.

pontusmelke committed:
Break out to one method per direction

Instead of sending in the direction and switch on the direction we have
separate `outgoing`, `incoming`, and `all` methods and leave it to the
caller to choose cursor/iterator.
